running workload/schemachange causes 19.2 node in mixed version cluster to crash

**Describe the problem**

When running the following workload on a mixed version cluster that has been upgraded to `20.1` and then downgraded (1 node) to `19.2.5`. The downgraded node crashes.

**To Reproduce**

1. Start a 3-node `19.2.5` cluster.
Note: prevent the cluster from finalizing the upgrade by:
`SET CLUSTER SETTING cluster.preserve_downgrade_option = '19.2';

2. Upgrade each node to `20.1`.

3. Downgrade node 3 to `19.2.5`.

4. Run this workload:
`roachprod run $CLUSTER:4 -- ./workload run schemachange --init --concurrency=2 --verbose=true --max-ops=300 "{pgurl:1-3}"`

5. Node 3 crashes with
```
E200407 02:02:45.722783 4883 sql/conn_executor.go:717  [n3,client=10.142.0.41:45994,user=root] a SQL panic has occurred while executing "SELECT column_name FROM [SHOW COLUMNS FROM table488] ORDER BY random() LIMIT 1": unexpected Family: 19
E200407 02:02:45.722870 4883 util/log/crash_reporting.go:207  [n3,client=10.142.0.41:45994,user=root] a panic has occurred!
panic: unexpected Family: 19 [recovered]
	panic: panic while executing 1 statements: SELECT _ FROM [SHOW COLUMNS FROM _] ORDER BY random() LIMIT _; caused by unexpected Family: 19
```

I will be adding more details tomorrow.

--- a/docs/generated/settings/settings.html
+++ b/docs/generated/settings/settings.html
@@ -68,6 +68,6 @@
 <tr><td><code>trace.debug.enable</code></td><td>boolean</td><td><code>false</code></td><td>if set, traces for recent requests can be seen in the /debug page</td></tr>
 <tr><td><code>trace.lightstep.token</code></td><td>string</td><td><code></code></td><td>if set, traces go to Lightstep using this token</td></tr>
 <tr><td><code>trace.zipkin.collector</code></td><td>string</td><td><code></code></td><td>if set, traces go to the given Zipkin instance (example: '127.0.0.1:9411'); ignored if trace.lightstep.token is set</td></tr>
-<tr><td><code>version</code></td><td>custom validation</td><td><code>19.2-16</code></td><td>set the active cluster version in the format '<major>.<minor>'</td></tr>
+<tr><td><code>version</code></td><td>custom validation</td><td><code>19.2-17</code></td><td>set the active cluster version in the format '<major>.<minor>'</td></tr>
 </tbody>
 </table>

--- a/pkg/clusterversion/cockroach_versions.go
+++ b/pkg/clusterversion/cockroach_versions.go
@@ -57,6 +57,7 @@ const (
 	VersionStatementDiagnosticsSystemTables
 	VersionSchemaChangeJob
 	VersionSavepoints
+	VersionTimeTZType
 
 	// Add new versions here (step one of two).
 )
@@ -443,6 +444,11 @@ var versionsSingleton = keyedVersions([]keyedVersion{
 		// records.
 		Key:     VersionSavepoints,
 		Version: roachpb.Version{Major: 19, Minor: 2, Unstable: 16},
+	},
+	{
+		// VersionTimeTZType enables the use of the TimeTZ data type.
+		Key:     VersionTimeTZType,
+		Version: roachpb.Version{Major: 19, Minor: 2, Unstable: 17},
 	},
 	// Add new versions here (step two of two).
 

--- a/pkg/clusterversion/versionkey_string.go
+++ b/pkg/clusterversion/versionkey_string.go
@@ -33,11 +33,12 @@ func _() {
 	_ = x[VersionStatementDiagnosticsSystemTables-22]
 	_ = x[VersionSchemaChangeJob-23]
 	_ = x[VersionSavepoints-24]
+	_ = x[VersionTimeTZType-25]
 }
 
-const _VersionKey_name = "Version19_1VersionStart19_2VersionLearnerReplicasVersionTopLevelForeignKeysVersionAtomicChangeReplicasTriggerVersionAtomicChangeReplicasVersionTableDescModificationTimeFromMVCCVersionPartitionedBackupVersion19_2VersionStart20_1VersionContainsEstimatesCounterVersionChangeReplicasDemotionVersionSecondaryIndexColumnFamiliesVersionNamespaceTableWithSchemasVersionProtectedTimestampsVersionPrimaryKeyChangesVersionAuthLocalAndTrustRejectMethodsVersionPrimaryKeyColumnsOutOfFamilyZeroVersionRootPasswordVersionNoExplicitForeignKeyIndexIDsVersionHashShardedIndexesVersionCreateRolePrivilegeVersionStatementDiagnosticsSystemTablesVersionSchemaChangeJobVersionSavepoints"
+const _VersionKey_name = "Version19_1VersionStart19_2VersionLearnerReplicasVersionTopLevelForeignKeysVersionAtomicChangeReplicasTriggerVersionAtomicChangeReplicasVersionTableDescModificationTimeFromMVCCVersionPartitionedBackupVersion19_2VersionStart20_1VersionContainsEstimatesCounterVersionChangeReplicasDemotionVersionSecondaryIndexColumnFamiliesVersionNamespaceTableWithSchemasVersionProtectedTimestampsVersionPrimaryKeyChangesVersionAuthLocalAndTrustRejectMethodsVersionPrimaryKeyColumnsOutOfFamilyZeroVersionRootPasswordVersionNoExplicitForeignKeyIndexIDsVersionHashShardedIndexesVersionCreateRolePrivilegeVersionStatementDiagnosticsSystemTablesVersionSchemaChangeJobVersionSavepointsVersionTimeTZType"
 
-var _VersionKey_index = [...]uint16{0, 11, 27, 49, 75, 109, 136, 176, 200, 211, 227, 258, 287, 322, 354, 380, 404, 441, 480, 499, 534, 559, 585, 624, 646, 663}
+var _VersionKey_index = [...]uint16{0, 11, 27, 49, 75, 109, 136, 176, 200, 211, 227, 258, 287, 322, 354, 380, 404, 441, 480, 499, 534, 559, 585, 624, 646, 663, 680}
 
 func (i VersionKey) String() string {
 	if i < 0 || i >= VersionKey(len(_VersionKey_index)-1) {

--- a/pkg/sql/alter_table.go
+++ b/pkg/sql/alter_table.go
@@ -144,6 +144,14 @@ func (n *alterTableNode) startExec(params runParams) error {
 				return unimplemented.NewWithIssue(32917,
 					"adding a REFERENCES constraint while also adding a column via ALTER not supported")
 			}
+			version := params.ExecCfg().Settings.Version.ActiveVersionOrEmpty(params.ctx)
+			if !isTypeSupportedInVersion(version, d.Type) {
+				return pgerror.Newf(
+					pgcode.FeatureNotSupported,
+					"type %s is not supported until version upgrade is finalized",
+					d.Type.SQLString(),
+				)
+			}
 
 			newDef, seqDbDesc, seqName, seqOpts, err := params.p.processSerialInColumnDef(params.ctx, d, tn)
 			if err != nil {
@@ -869,6 +877,15 @@ func applyColumnMutation(
 	switch t := mut.(type) {
 	case *tree.AlterTableAlterColumnType:
 		typ := t.ToType
+
+		version := params.ExecCfg().Settings.Version.ActiveVersionOrEmpty(params.ctx)
+		if !isTypeSupportedInVersion(version, typ) {
+			return pgerror.Newf(
+				pgcode.FeatureNotSupported,
+				"type %s is not supported until version upgrade is finalized",
+				typ.SQLString(),
+			)
+		}
 
 		// Special handling for STRING COLLATE xy to verify that we recognize the language.
 		if t.Collation != "" {

--- a/pkg/sql/create_table.go
+++ b/pkg/sql/create_table.go
@@ -112,6 +112,21 @@ var storageParamExpectedTypes = map[string]storageParamType{
 	`user_catalog_table`:                          storageParamUnimplemented,
 }
 
+// minimumTypeUsageVersions defines the minimum version needed for a new
+// data type.
+var minimumTypeUsageVersions = map[types.Family]clusterversion.VersionKey{
+	types.TimeTZFamily: clusterversion.VersionTimeTZType,
+}
+
+// isTypeSupportedInVersion returns whether a given type is supported in the given version.
+func isTypeSupportedInVersion(v clusterversion.ClusterVersion, t *types.T) bool {
+	minVersion, ok := minimumTypeUsageVersions[t.Family()]
+	if !ok {
+		return true
+	}
+	return v.IsActive(minVersion)
+}
+
 // ReadingOwnWrites implements the planNodeReadingOwnWrites interface.
 // This is because CREATE TABLE performs multiple KV operations on descriptors
 // and expects to see its own writes.
@@ -1237,6 +1252,13 @@ func MakeTableDesc(
 						"VECTOR column types are unsupported",
 					)
 				}
+			}
+			if !isTypeSupportedInVersion(version, d.Type) {
+				return desc, pgerror.Newf(
+					pgcode.FeatureNotSupported,
+					"type %s is not supported until version upgrade is finalized",
+					d.Type.SQLString(),
+				)
 			}
 			if d.PrimaryKey.Sharded {
 				// This function can sometimes be called when `st` is nil,

--- a/pkg/sql/logictest/testdata/logic_test/alter_table_mixed_19.2_20.1
+++ b/pkg/sql/logictest/testdata/logic_test/alter_table_mixed_19.2_20.1
@@ -159,3 +159,17 @@ ALTER TABLE t RENAME COLUMN b TO c
 
 statement error schema change cannot be initiated in this version until the version upgrade is finalized
 ALTER TABLE t RENAME TO s
+
+subtest regression_47110
+
+statement error type TIMETZ is not supported until version upgrade is finalized
+CREATE TABLE regression_47110(a TIMETZ)
+
+statement ok
+CREATE TABLE regression_47110(a TIME)
+
+statement error type TIMETZ is not supported until version upgrade is finalized
+ALTER TABLE regression_47110 ADD COLUMN b TIMETZ
+
+statement error type TIMETZ is not supported until version upgrade is finalized
+ALTER TABLE regression_47110 ALTER a SET DATA TYPE timetz

